### PR TITLE
Added Docker setup.

### LINKED ISSUE #80: Dockerfile
Originally reported by: **BitBucket: [almereyda](http://bitbucket.org/almereyda), GitHub: [almereyda](http://github.com/almereyda)**

---

As I would love to see uMap spreading across the world, let's start talking about a Dockerfile here.
- Which prequisites do we need?
- Which succsessful (Geo)Django Dockerfiles are known, what can we build upon?
  - https://registry.hub.docker.com/search?q=django
  - https://github.com/dockerfile?tab=repositories

It is to mention, that Docker setups prefer stateless apps, i.e. decoupled logic and persistence. http://12factor.net/ gives an introduction to that thought.

---

Once I've got that working, I'd like to provide a GeoCouch adapter for Django and even further seperate the UI from the backend, so it could run as a CouchApp.

---
- Bitbucket: https://bitbucket.org/yohanboniface/umap/issue/80

### PR DESCRIPTION
**Pull request by jezdez**
This adds:

- a Dockerfile
- a Docker compose file for easy testing
- a Travis CI setup
  - so that it can build a Docker image and push to Docker Hub automatically
  - it does that on every Git tag as well and push a equally tagged version to Docker Hub
- extends the Makefile to add some helper tasks for docker (e.g. make docker-test)

Fix #80

### CODE REVIEW

**Comment by jezdez** [comment]
@yohanboniface 

- [ ] Could you please take a look at the settings file and if that's the env var names you like? Any settings missing?

Also, there are a few assumptions I've made:

- [ ] you'd have to enable Travis CI for the repo: https://travis-ci.org/

- [ ] sign up for Docker Hub and let me know the username so I can set the `DOCKER_IMAGE_NAME` env var name in the `.travis.yml` accordingly, as it consists of the username and project you chose for uMap

- [ ] in addition you'll have to provide your Docker username and password to the [Travis CI project settings ](https://docs.travis-ci.com/user/environment-variables/#Defining-Variables-in-Repository-Settings) as `DOCKER_USERNAME` and `DOCKER_PASSWORD`.

File: umap/settings/docker.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+Settings for Docker development
+
+Use this file as a base for your local development settings and copy
+it to umap/settings/local.py. It should not be checked into
+your code repository.
+"""
+import environ
+from umap.settings.base import *   # pylint: disable=W0614,W0401
+
+env = environ.Env()
```
Comment (yohanboniface) [inline umap/settings/docker.py]: I'd suggest to keep in `docker.py` only the specific bits, so I'd move the env sniffing in `base.py`. Appart from the new depss, it does not hurt, and may even save someone's day :)

**Comment by yohanboniface** [comment]
> sign up for Docker Hub and let me know the username so I can set the DOCKER_IMAGE_NAME env var name in the .travis.yml accordingly, as it consists of the username and project name you chose for uMap

My username is "yohanboniface". I've also added "umap" as organisation, and you as this organisation maintainer.

> in addition you'll have to provide your Docker username and password to the Travis CI project settings as DOCKER_USERNAME and DOCKER_PASSWORD (don't paste them into .travis.yml)

No way to add in encrypted in the travis.yml?

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+language: python
+
+python: 3.5
+
+services:
+  - docker
+
+install:
+  - pip install -e .[dev]
```
Comment (yohanboniface) [inline .travis.yml]: Shouldn't we create the psql db to be able to run tests?

File: umap/settings/docker.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+SOCIAL_AUTH_TWITTER_KEY = env('TWITTER_KEY', default='')
+SOCIAL_AUTH_TWITTER_SECRET = env('TWITTER_SECRET', default='')
+if SOCIAL_AUTH_TWITTER_KEY and SOCIAL_AUTH_TWITTER_SECRET:
+    AUTHENTICATION_BACKENDS += (
+        'social_core.backends.twitter.TwitterOAuth',
+    )
+SOCIAL_AUTH_OPENSTREETMAP_KEY = env('OPENSTREETMAP_KEY', default='')
+SOCIAL_AUTH_OPENSTREETMAP_SECRET = env('OPENSTREETMAP_SECRET', default='')
+if SOCIAL_AUTH_OPENSTREETMAP_KEY and SOCIAL_AUTH_OPENSTREETMAP_SECRET:
```
Comment (yohanboniface) [inline umap/settings/docker.py]: Out of scope for this PR, but someone may want to run uMap with another OAuth provider. Not sure how to make this easy/possible with Docker.

**Comment by ixeft** [comment]
Hello,
What is the status of this PR ? 
I would be interested in this functionality, and it seems quit advanced :)

**Comment by jezdez** [comment]
Ugh, I forgot I did this, not sure. @yohanboniface Do you still want/need this?

**Comment by yohanboniface** [comment]
A uMap Docker image would of course be awesome!

But I'm missing Docker habits myself and time to join the effort. Mainly, I've bad (old) experience with data management when updating a docker image.

I'll try to have a deeper look anyway.

**Comment by xthursdayx** [comment]
I'm super interested in this as well. I had tried to figure out how to create a docker container of uMap myself, but the limits of my technical capacity quickly became apparent.

**Comment by drunkod** [comment]
I support it.
  It speeds up development!

**Comment by fti7** [comment]
Please merge! :-)

**Comment by lyz-code** [comment]
@yohanboniface @jezdez Any advance on this PR? 

If you specify what does it lack to be able to be merged into master I may be able to help.

**Comment by jezdez** [comment]
No idea, I’m sorry, I haven’t had time to look into it and don’t have a need for it anymore.

**Comment by Duvel** [comment]
Maybe it could be an idea to split the travis part from the docker part and add that later?

@yohanboniface did add some comments, but I think that these also can be resolved later.

One of the things currently missing in the docker container is correct logging. I have version 0.8.2 running in docker, but version 1.1.2 is giving errors and I have no clue why, because of missing logging.

**Comment by Duvel** [comment]
I got logging working in my test now, thanks to docker.py in msgre/uwsgi_logging@4055e52.

So probably that is something we would want to add.

**Comment by Fi1osof** [comment]
Did try this. Many errors. Sad :(
@jezdez, thanks for trying!

**Comment by Duvel** [comment]
@Fi1osof I updated my fork with the files, that I used to get a working docker image. It is based on the work done by @jezdez and I updated it for the newer versions of uMap. I added a pull request #801 if we would like to have this as part of this project.

**Comment by jezdez** [comment]
Closing in favor of #801.

**Pull request closed without merging**